audio/wav.go

```diff
@@ -0,0 +1,276 @@
+// information is used to create the WAV format struct
+func NewWAVFromData(data []byte) (*WAV, error) {
+	// create a WAV from the given buffer.
+	// return error if len(data) < 44
+	// extract data from the data according to the spec: http://soundfile.sapp.org/doc/WaveFormat/
+
+	// find first data index
+	i := 4
+	for i < len(data) && data[i-4] != 'd' || data[i-3] != 'a' || data[i-2] != 't' || data[i-1] != 'a' {
+		i++
+	}
+
+	dataLen := len(data) - i
+	if dataLen <= 0 {
+		return nil, errors.Error{
+			Code:    "One",
+			Message: "Received WAV file with empty data",
+		}
+	}
+
+	// hOff is the header offset. It denotes the start of the header, which is
+	// 44 bytes before the data starts.
+	hOff := i - 44
+
+	numChannels := binary.LittleEndian.Uint16(data[hOff+22 : hOff+23])
+	sampleRate := binary.LittleEndian.Uint32(data[hOff+24 : hOff+27])
+	bitsPerSample := binary.LittleEndian.Uint16(data[hOff+34 : hOff+35])
+	audioData := data[i:]
+
+	return &WAV{
+		NumChannels:   numChannels,
+		SampleRate:    sampleRate,
+		AudioFormat:   DefaultAudioFormat,
+		BitsPerSample: bitsPerSample,
+		audioData:     audioData,
+	}, nil
+}
+
+// NewWAVFromReader takes in a reader and creates a new WAV format 
+// with the given information. 
+func NewWAVFromReader(reader io.Reader) (*WAV, error) {
+	b, err := ioutil.ReadAll(reader)
+	if err != nil {
+		return nil, err
+	}
+	return NewWAVFromData(b)
+}
+
+// getRMS is a helper function used to calculate the RMS. This is called 
+// by TrimSilent which uses RMS to determine whether the sample of audio
+// is silent
+func getRMS(sampleSize uint16, audioData []byte) float64 {
+	sum := 0.0
+	for i := uint16(0); i < uint16(len(audioData)); i += sampleSize {
+		val := binary.LittleEndian.Uint64(audioData[i:(i + sampleSize)])
+		sum += math.Pow(2.0, float64(val))
+	}
+
+	return math.Sqrt(sum / (float64(len(audioData)/int(sampleSize))))
+}
+
+// TrimSilent is called on a WAV struct to trim the silent portions from the
+// ends of the file while leaving a certain amount of padding
+func (w *WAV) TrimSilent(threshold float64, padding float64) *WAV {
+	// trim silence from the ends of the file, leaving a certain amount of padding
+	// numSamples := uint16(len(w.audioData)) * 8 / uint16(w.BitsPerSample)
+	sizeOfSample := uint16(w.BitsPerSample / 8)
+	rmsSampleCount := uint16((1.0 / 16.0) * float64(w.SampleRate))
+
+	// get max amplitude
+	maxPossibleAmp := math.Exp2(float64(w.BitsPerSample))
+	maxPossibleVal := maxPossibleAmp / 2.0
+
+	silenceThresh := threshold * maxPossibleVal
+
+	// Trimming the beginning
+	N1 := uint16(0)
+
+	for N1 < uint16(len(w.audioData)) {
+		// sampleValue := binary.LittleEndian.Uint16(w.audioData[N1:N1+sizeOfSample-1])
+		// sumSquares += math.Pow(float64(2), float64(sampleValue))
+		// rms := math.Sqrt(sumSquares / float64(numSamples))
+		rms := getRMS(sizeOfSample, w.audioData[N1:N1+sizeOfSample*rmsSampleCount])
+
+		N1 += sizeOfSample * rmsSampleCount
```

Comment:
don't add this if the audio section wasn't silent. This will result in chopping off some audio. Also, I don't think the value of `rmsSampleCount` was calculated correctly if its being used in this context.